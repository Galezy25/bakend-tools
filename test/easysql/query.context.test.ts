import QueryContext from '../../src/easysql/query.context';

describe('QueryContext', () => {
    test('Field filters', () => {
        let queryContext = new QueryContext()
            .field('id')
            .notEqualTo('1234567890')
            .field('date')
            .between('2025/01/01 00:00:00', '2025/07/01 00:00:00')
            .field('price')
            .lessOrEqualThan(200)
            .field('price')
            .greaterThan(10);
        expect(queryContext.getPureContext()).toMatchObject({
            id_n: '1234567890',
            date_btw: '2025/01/01 00:00:00,2025/07/01 00:00:00',
            price_lseq: 200,
            price_gt: 10,
        });
        expect(queryContext.getFilterSentences()).toMatchSnapshot();
        queryContext.clearContext();

        queryContext
            .field('id')
            .notEqualTo(['789123456', '456789123'])
            .field('id')
            .like(['7%', '4%'])
            .field('price')
            .greaterOrEqualThan(10)
            .field('price')
            .lessThan(200);
        expect(queryContext.getPureContext()).toMatchObject({
            id_n: ['789123456', '456789123'],
            id_like: ['7%', '4%'],
            price_ls: 200,
            price_gteq: 10,
        });
        expect(queryContext.getFilterSentences()).toMatchSnapshot();
    });

    test('Table related', () => {
        let ctxWithRelations = new QueryContext(['users']);
        let ctxWithNoRelations = new QueryContext(undefined,undefined);

        expect(
            ctxWithRelations.populate(
                'left',
                'users',
                'this.leader:users.id',
                'name:leader_name'
            )
        ).toBeDefined();
        expect(ctxWithRelations.getPureContext()).toMatchObject({
            users_left: 'this.leader:users.id,name:leader_name',
        });
        expect(ctxWithRelations.getRelations()).toMatchSnapshot();
        expect(ctxWithRelations.getSelectedFields()).toMatchSnapshot();
        ctxWithRelations.fields('id', 'name', 'email');
        expect(ctxWithRelations.getSelectedFields()).toMatchSnapshot();
        expect(() =>
            ctxWithNoRelations.populate(
                'left',
                'users',
                'this.leader:users.id',
                'users.name:leader_name'
            )
        ).toThrowError();
        expect(ctxWithNoRelations.getPureContext()).not.toMatchObject({
            users_left: 'this.leader:users.id,users.name:leader_name',
        });
        expect(ctxWithNoRelations.getRelations()).toBe('');
        expect(ctxWithNoRelations.getSelectedFields()).toBe('this.*');
    });

    test('setPureContext', () => {
        let ctxSaleProducts = new QueryContext(['products']).setPureContext({
            products_inner:
                'this.product:products.id,description:product_description',
            _fields: 'id,amount,unit_cost,product:product_id',
            client: ['123789', '456127'],
            date_btw: '2025/01/01 00:00:00,2025/07/01 00:00:00',
            sale_nin: '1834,7890',
            _sort: 'sale:ASC',
            _limit: 250,
        });
        expect(ctxSaleProducts.getSelectedFields()).toMatchSnapshot();
        expect(ctxSaleProducts.getRelations()).toMatchSnapshot();
        expect(ctxSaleProducts.getFilterSentences()).toMatchSnapshot();
        expect(ctxSaleProducts.getOrderSentence()).toMatchSnapshot();
        expect(ctxSaleProducts.getLimitSentence()).toMatchSnapshot();

        let ctxToClone = new QueryContext()
            .field('id')
            .notEqualTo('1234567890')
            .field('date')
            .between('2025/01/01 00:00:00', '2025/07/01 00:00:00')
            .field('price')
            .lessOrEqualThan(200)
            .field('price')
            .greaterThan(10);
        let ctxCloned = new QueryContext().setPureContext(
            ctxToClone.getPureContext()
        );
        expect(ctxToClone.getPureContext()).toMatchObject(
            ctxCloned.getPureContext()
        );
    });
});
