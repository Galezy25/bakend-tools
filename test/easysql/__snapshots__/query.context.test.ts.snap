// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryContext Field filters 1`] = `"id <> '1234567890' AND date BETWEEN '2025/01/01 00:00:00' AND '2025/07/01 00:00:00' AND price <= 200 AND price > 10"`;

exports[`QueryContext Field filters 2`] = `"(id <> '789123456' OR id <> '456789123') AND (id LIKE '7%' OR id LIKE '4%') AND price >= 10 AND price < 200"`;

exports[`QueryContext Table related 1`] = `"LEFT JOIN users ON this.leader = users.id"`;

exports[`QueryContext Table related 2`] = `"this.*, users.name AS leader_name"`;

exports[`QueryContext Table related 3`] = `"this.id, this.name, this.email, users.name AS leader_name"`;

exports[`QueryContext setPureContext 1`] = `"this.id, this.amount, this.unit_cost, this.product AS product_id, products.description AS product_description"`;

exports[`QueryContext setPureContext 2`] = `"INNER JOIN products ON this.product = products.id"`;

exports[`QueryContext setPureContext 3`] = `"(client = '123789' OR client = '456127') AND date BETWEEN '2025/01/01 00:00:00' AND '2025/07/01 00:00:00' AND sale NOT IN ('1834','7890')"`;

exports[`QueryContext setPureContext 4`] = `"ORDER BY sale ASC"`;

exports[`QueryContext setPureContext 5`] = `"LIMIT 250"`;
